Unicode character support in screen tab names

Explorer09 opened the pull request:
I would like to share a feature that I have done as part of a code refactoring process.

These changes are still work in progress, but I think it's good to file it and gather early feedback.

This feature makes the screen tab names display with the correct terminal widths when they contain Unicode characters (in UTF-8 encoding). It replaces the naive `strlen` calculation with a new function that calculates the display terminal width of a name string.

The PR introduces the following functions:
* `String_makePrintable`, a sanitization function with an API similar to`xStrndup`. It reads the string, converts all unprintable characters to replacement code points (U+FFFD), and allocates and re-encodes the string.
* `EncodePrintableString`, an internal routine for `String_makePrintable`. This routine is designed with room to support `RichString` buffers
* `String_mbswidth`, a function that calculates the terminal width of a string, or calculate the byte length of a substring if it is clipped by a terminal width limit.

This PR  is related to #1513.

BenBE commented on XUtils.c:
```diff
@@ -224,6 +232,209 @@ size_t String_safeStrncpy(char* restrict dest, const char* restrict src, size_t
+      }
+      wasReplaced = shouldReplace;
+
+      destSize += encodeWChar((dest ? (char*)dest + destSize : NULL), wc, encState);
```
Treat the `destSize` as opaque, similar to how it is done in the CGroup compression code. The location needs to be part of the passed state, alongside the maximum buffer size. This avoids any pointer arithmetic on `dest` here and improves re-usability with other memory layouts.

Explorer09 replied:
I'm reserved about this. `destSize` is required to be returned to the caller so that the caller can `malloc` the buffer that just fits. It would not be that "opaque" as you think.

It might be a good idea for `EncodePrintableString` to not care about the `destSize` at all and have the property and the counting all done by the callee (`encodeWChar`). The caller of  `EncodePrintableString` would have to extract the  `destSize` property from `encState` nevertheless.

BenBE commented on XUtils.c:
```diff
@@ -224,6 +232,209 @@ size_t String_safeStrncpy(char* restrict dest, const char* restrict src, size_t
+   size_t destSize = 0;
+   bool wasReplaced = false;
+
+#ifdef HAVE_LIBNCURSESW
```
Any chance we could split this into two smaller blocks? Currently there is much code duplication.

Explorer09 replied:
I wished to do that, but the problem is on the data type. One is `char` and the other is `wchar_t`. How do I resolve this? (Introduce something like a `TCHAR` type in Windows programming?)

BenBE commented on XUtils.c:
```diff
@@ -224,6 +232,209 @@ size_t String_safeStrncpy(char* restrict dest, const char* restrict src, size_t
+   mbstate_t newState;
+   wchar_t wc;
+   do {
+      wc = L'\0';
```
```suggestion
      wc = 0;
```

This makes `c` and `wc` assignments compatible and thus would not require distinction in the code paths …

BenBE commented on XUtils.c:
```diff
@@ -224,6 +232,209 @@ size_t String_safeStrncpy(char* restrict dest, const char* restrict src, size_t
+
+#endif /* HAVE_LIBNCURSESW */
+
+   assert(destSize > 0);
```
We shouldn't care on this level about `destSize`. All we should return is the overall success of the conversion. Cf. CGroup compression code.

Explorer09 replied:
This function, `EncodePrintableString`, is intended to be called twice. One for counting the size needed and the second time for doing the actual conversion. There is no buffer expansion (a la `realloc`) in the middle of the process.

`EncodePrintableString` will return the size, and then the caller will allocate the buffer, and call the same function again to do the conversion.

fasterit commented:
Fails CI:
```
XUtils.c:426:6: error: 'HAVE_STRNLEN' is not defined, evaluates to 0 [-Werror,-Wundef]
  426 | # if HAVE_STRNLEN
      |      ^
1 error generated.``

Explorer09 commented:
> Fails CI:
> 
> ```
> XUtils.c:426:6: error: 'HAVE_STRNLEN' is not defined, evaluates to 0 [-Werror,-Wundef]
>   426 | # if HAVE_STRNLEN
>       |      ^
> 1 error generated.``
> ```

Noted. It should have been `#ifdef`.

Explorer09 commented on XUtils.c:
```diff
@@ -22,6 +23,13 @@ in the source distribution for its full text.
 #include "CRT.h"
 #include "Macros.h"
 
+#ifdef HAVE_LIBNCURSESW
```
The htop codebase doesn't support any curses**w** implementation other than `ncursesw` for now.

You are right about that I can simply include `ProvideCurses.h` instead.

BenBE replied:
The idea of `ProvideCurses.h` basically is to provide you with curses¹. So anything related there should follow along to minimize churn from messing with all the flags involved with it.

I even thought about having the `character type` like `wchar_t` be a define, where differences between enabled and disabled Unicode support arise. Though I dropped that idea …

¹pun intended …

BenBE commented:
> > Fails CI:
> > ```
> > XUtils.c:426:6: error: 'HAVE_STRNLEN' is not defined, evaluates to 0 [-Werror,-Wundef]
> >   426 | # if HAVE_STRNLEN
> >       |      ^
> > 1 error generated.``
> > ```
> 
> Noted. It should have been `#ifdef`.

IDK, but nowhere else do we check for `strnlen`, so why here all of a sudden?

BenBE commented:
Any objections to pull in 4ccfc38335aaaaad3126844b7046e310a7cf6d59 to `main` now and do the Unicode support later? That commit is mostly cleanup and doesn't affect much of the later work in this PR.

BenBE commented:
> > Any objections to pull in [4ccfc38](https://github.com/htop-dev/htop/commit/4ccfc38335aaaaad3126844b7046e310a7cf6d59) to `main` now and do the Unicode support later? That commit is mostly cleanup and doesn't affect much of the later work in this PR.
> 
> I have just one question with the commit: Is it okay for the constants to be defined in `CRT.h` or is there a better header to put those two constant names?

`CRT.h` seems to be fine. `ScreenManager.h` may be another place to put them, but actually I'm not too sure, where's best …

> BTW, I borrowed the "margin left" and "column gap" naming from CSS. 🙂

Fine.

BenBE commented on XUtils.c:
```diff
@@ -224,6 +226,185 @@ size_t String_safeStrncpy(char* restrict dest, const char* restrict src, size_t
+
+   mbstate_t decState;
+   memset(&decState, 0, sizeof(decState));
+   mbstate_t newState;
```
This can be put into the loop; no need to track it beyond loop iterations.

Explorer09 replied:
Need to test if this move has an impact on code size...

Explorer09 replied:
```diff
diff --git a/XUtils.c b/XUtils.c
index f841a104..767411ba 100644
--- a/XUtils.c
+++ b/XUtils.c
@@ -269,7 +269,6 @@ void EncodePrintableString(WCharEncoderState* ps, const char* src, size_t maxLen
 
    mbstate_t decState;
    memset(&decState, 0, sizeof(decState));
-   mbstate_t newState;
 #else
    const char replacementChar = '?';
    char ch;
@@ -283,6 +282,7 @@ void EncodePrintableString(WCharEncoderState* ps, const char* src, size_t maxLen
       if (pos < maxLen) {
          // Read the next character from the byte sequence
 #ifdef HAVE_LIBNCURSESW
+         mbstate_t newState;
          memcpy(&newState, &decState, sizeof(newState));
          len = mbrtowc(&ch, &src[pos], maxLen - pos, &newState);
 
```

Result. With the small change like this,
1. GCC 13.2 (Ubuntu 24.04 "Noble") shows no code size changes.
2. Clang 18 (Ubuntu 22.04 "Jammy"): code size becomes 7 bytes bigger with `-Os` option.
3. Clang 20 (Ubuntu 24.04 "Noble"): code size becomes 27 bytes smaller with `-Os` option.

So it's interesting to see Clang missed this. The code should be equivalent and yet Clang generated the binaries with different sizes.

BenBE commented on XUtils.c:
```diff
@@ -224,6 +226,185 @@ size_t String_safeStrncpy(char* restrict dest, const char* restrict src, size_t
+         len = mbrtowc(&ch, &src[pos], maxLen - pos, &newState);
+
+         switch (len) {
+
```
No blank line needed.
```suggestion
```

BenBE commented on XUtils.h:
```diff
@@ -23,8 +23,21 @@ in the source distribution for its full text.
 #include "Macros.h"
+#include "ProvideCurses.h"
 
 
+typedef struct WCharEncoderState_ {
+   size_t pos;
+   void* buf;
+   mbstate_t mbState;
+} WCharEncoderState;
```
This should also carry the buffer size, so the internal `EncodeWChar` routines can perform bound checking.

Also note, that the CGroup code deliberately uses two distinct functions (one for counting the size, another for actually writing) and that these functions return success (character properly counted/written to buffer) or failure (OOB write). Similar could be done here too.

Explorer09 replied:
I choose to make the size counting and the actual writing the same routine for a reason: I expect the "rules" for filtering characters be tweakable in the future. If there are additional characters besides `!iswprint` that need to be filtered out (replaced with U+FFFD), people only need to modify one routine for that.

I'm no sure about bound checking yet. I can add it into the struct, but it would be for assertions only, as I always expect the buffer allocated would just fit. (That is, no truncation.)

BenBE replied:
That's what "Failure to write to the buffer" could be used for. If the `EncodeWChar` function sees that the buffer is full and not NUL-terminated, it should return `false` (AKA failure), which all callees could propagate accordingly. Similar to what the CGROUP compression code¹ does …

¹Actually that code over-allocates by one byte and forces the NUL-termination in that over-allocated byte. Cf. `linux/CGroupUtils.c` in functions `CGroup_filterName` and `CGroup_filterContainer`.

Explorer09 replied:
@BenBE Nope. After a bit of consideration, this length check cannot be done in `EncodeWChar` like in the CGROUP code. I wrote a comment on why instead.

It's a limitation of the `wcrtomb(3)` interface. When we can detect `wcrtomb` writing out of bounds we might be too late (access violation already happens). `wcrtomb` doesn't give us the "length check before writing" we need. Trying to workaround that would make code unnecessarily complicated (`wcrtomb` called one more time - three times in total).

Also, we cannot put the `'\0'` byte in arbitrary position. Unlike in the CGROUP code, we can only put `'\0'` in character boundaries here. So overall there's nothing I can do.

Explorer09 replied:
@BenBE Update: This is what I mean by being "unnecessarily expensive".

```c
static bool String_encodeWChar(WCharEncoderState* ps, wchar_t wc) {
   assert(!ps->buf || ps->pos < ps->size);
   
   char tempBuf[MB_LEN_MAX];
   size_t len = wcrtomb(tempBuf, wc, &ps->mbState);
   if (len == (size_t)-1) {
      return false;
   }
   assert(len > 0);
   if (ps->buf) {
      if (len > ps->size - ps->pos) {
         // When this happens, ps->buf will not be null-terminated and
         // ps->mbState becomes undefined state (cannot be reused).
         // Thus we have no way to recover from here.
         return false;
      }
      char* buf = ps->buf;
      memcpy(&buf[ps->pos], tempBuf, len);
   }

   ps->pos += len;
   return true;
}
```

BenBE replied:
Yes. And with the implementation used in the CGroup code, a `false` returned from the buffer write callback would immediately terminate processing. That's basically the same semantics used there too.

Explorer09 replied:
@BenBE That means an additional `memcpy` call, slowing things down slightly. For checking an error that should never happen unless there's a bug in code. I don't really like this.

BenBE replied:
That `memcpy` is from L1 cache to a buffer that's likely in L1 too. As mentioned elsewhere: If in doubt, I prefer correctness over performance.

And if you want to avoid this, over-allocate by `MB_LEN_MAX + 1` and you can always write to `ps->buf`, if you force the NUL terminator on error.

Explorer09 replied:
> And if you want to avoid this, over-allocate by `MB_LEN_MAX + 1` and you can always write to `ps->buf`, if you force the NUL terminator on error.

_Over-allocation_ to solve the (actually non-existent) problem of _under-allocation_. Sorry, but I'm irritated by this logic.

The whole point of adding the bound check by your suggestion was to correct the case of _under-allocation_ (allocated buffer size is less than required). While I argued first that the problem literally won't exist, I agree on the point that we can _assert_ and _detect_ the error, at least as a precaution. The buffer should be allocated with the right size already, thus any code that attempts to _correct_ or _recover_ from any potential error is useless. I might just make the error fatal instead and call it a day.

```c
if (ps->buf && len <= ps->size - ps->pos) {
   fail();
}
```

As for why there is no NUL termination for the string when odds happen here: We are using a multibyte encoder `wcrtomb` that is capable of handling any encoding, not just UTF-8. When interpreting the standard behavior of `wcrtomb` in the strict sense, _it is not necessary for a null wide character to be encoded as 1 byte_. The standard `wcrtomb` will output shift sequences (if any) that restore the shift state to the initial state before outputting the null byte itself. Since that would require extra buffer space (up to `MB_CUR_MAX` bytes), any truncation of a multibyte string would have to be decided beforehand, and it won't be an escape hatch for any buffer size problem.

BenBE replied:
ACK on making the error fatal.

Anyway, a good API makes it impossible to cause issues by abusing it. Thus if `wcrtomb` may write up to `MB_LEN_MAX` bytes you better should pass a buffer that can hold `MB_LEN_MAX` bytes, i.e. ensure the remaining space is large enough. Even if the actual call is known to write less.

And adding `MB_LEN_MAX + 1` to the allocation won't cause any noticeable overhead (malloc will just allocate the same and in rare occasions or under memory pressure may delay due to the OOM killer sweeping some swap …

BenBE commented on configure.ac:
```diff
@@ -372,6 +372,7 @@ AC_CHECK_FUNCS([ \
     sched_getscheduler \
     sched_setscheduler \
     strchrnul \
+    strnlen \
```
I'm not sure I like this, when we already have plenty of code, that requires this function without prior check …

Explorer09 replied:
A quick `grep -r strnlen` shows no code in htop has been using `strnlen` before this PR.

Explorer09 replied:
Update. `strnlen` was recently added in #1707.

BenBE commented on XUtils.c:
```diff
@@ -224,6 +226,185 @@ size_t String_safeStrncpy(char* restrict dest, const char* restrict src, size_t
+}
+#else
+static void String_encodeWChar(WCharEncoderState* ps, int c) {
+   if (ps->buf) {
+      ((char*)ps->buf)[ps->pos] = (char)c;
+   }
```
```suggestion
   char* buf = ps->buf;
   if (buf) {
      buf[ps->pos] = (char)c;
   }
```

BenBE commented on XUtils.c:
```diff
@@ -224,6 +226,185 @@ size_t String_safeStrncpy(char* restrict dest, const char* restrict src, size_t
+   encState.buf = buf;
+   EncodePrintableString(&encState, str, maxLen, String_encodeWChar);
+   assert(encState.pos == bufSize);
+   encState.buf = NULL;
```
Why the explicit `NULL` assignment?

Explorer09 replied:
I thought Clang's analyzer would be confused, thinking `encState.buf` being the owner of the allocated buffer. It turns out it wasn't confused here and so I removed the line.

BenBE commented on XUtils.c:
```diff
@@ -224,6 +226,185 @@ size_t String_safeStrncpy(char* restrict dest, const char* restrict src, size_t
+   char* buf = xMalloc(bufSize);
+   encState.buf = buf;
+   EncodePrintableString(&encState, str, maxLen, String_encodeWChar);
+   assert(encState.pos == bufSize);
```
Technically the assertion should be:
```suggestion
   assert(encState.pos <= bufSize);
```

Practically, the encoding of the string itself should match what we counted earlier.

Explorer09 replied:
I don't expect oversized allocation here. I intended the size reported be equal. Not more or less.

BenBE commented on XUtils.c:
```diff
@@ -224,6 +226,185 @@ size_t String_safeStrncpy(char* restrict dest, const char* restrict src, size_t
+
+#ifndef HAVE_STRNLEN
+static size_t strnlen(const char* str, size_t maxLen) {
+   size_t len;
+   for (len = 0; len < maxLen && str[len] != '\0'; len++) {}
+   return len;
```
```suggestion
   for (size_t len = 0; len < maxLen; len++)
      if (!str[len])
         return len;
   return maxLen;
```

BenBE commented on XUtils.c:
```diff
@@ -224,6 +226,187 @@ size_t String_safeStrncpy(char* restrict dest, const char* restrict src, size_t
+}
+#else
+static void String_encodeWChar(WCharEncoderState* ps, int c) {
+   char* buf = (char*)ps->buf;
```
No need to cast `void*` to `char*` …

BenBE commented on Action.c:
```diff
@@ -432,7 +432,8 @@ Htop_Reaction Action_setScreenTab(State* st, int x) {
    for (unsigned int i = 0; i < settings->nScreens; i++) {
       const char* tab = settings->screens[i]->heading;
-      int width = rem >= bracketWidth ? (int)strnlen(tab, rem - bracketWidth + 1) : 0;
+      const char* ptr = tab;
```
Why introduce this extra variable? Was this patch written by AI? ;-)

Explorer09 replied:
Technically no AI is used :)

The `ptr` is a temporary variable holding a pointer because of the API design I explained in https://github.com/htop-dev/htop/pull/1642/changes/BASE..6b82bcef3b8f7aecd9be6c83d66e84af152a1d58#r3211053429

BenBE commented on XUtils.c:
```diff
@@ -259,6 +260,299 @@ size_t strnlen(const char* str, size_t maxLen) {
+   // It is unnecessarily expensive to fix the output string if the caller
+   // gives an incorrect buffer size. This function would not support any
+   // truncation of the output string.
+   size_t len = wcrtomb(dest, wc, &ps->mbState);
```
Potential for buffer overruns, if a position fewer than `MB_LEN_MAX` from the end of the buffer is passed. A safer option (which should hit L1) would be to always pass `tempBuf` and copy `len` bytes on success, unless that would write to `ps->buf[ps->size - 1]` (which should be forced `0`).

Explorer09 replied:
When I look at this code again I think your suggestion of always passing `tempBuf` to `wcrtomb` is the right choice. Not only it mitigates the buffer overrun problem, but it also prevents an undefined behavior when the input `wc` is an invalid code point.

Explorer09 replied:
Update: I've modified `String_encodeWChar` that its internal `wcrtomb` call now always writes to `tempBuf`. The function now never attempts to write out of bounds (regarding `ps->buf` and `ps->size`), however when the size is too small the error is still fatal. I don't think truncation of a string is an intended use case of this function, or forcing the output string to be null terminated (the string will be null terminated only upon the `String_encodeWChar(ps, '\0')` call).

BenBE commented on XUtils.c:
```diff
@@ -259,6 +260,299 @@ size_t strnlen(const char* str, size_t maxLen) {
+      fail();
+   }
+   if (ps->buf && len > ps->size - ps->pos) {
+      assert(!ps->buf || len <= ps->size - ps->pos);
```
This check is after an OOB write by `wcrtomb` already corrupted memory. This just cures the system if it survives til here. ASAN should complain inside the `wcrtomb` call above.

Explorer09 replied:
Applied the suggestion in [this comment](https://github.com/htop-dev/htop/pull/1642#discussion_r3210999167) and now the function no longer writes out of bounds.

BenBE commented on XUtils.c:
```diff
@@ -259,6 +260,299 @@ size_t strnlen(const char* str, size_t maxLen) {
+   return buf;
+}
+
+bool String_decodeNextWChar(MBStringDecoderState* ps) {
```
Naming consistency `String_encodeWChar` vs. `String_decodeNextWChar`.

Also, any reason to have this function exported from this TU?

Explorer09 replied:
There's a reason for asymmetric naming here. (You may suggest a better naming.)

`encodeWChar()` and `decodeNextWChar()` are for low-level encoder and decoder with states.

For the decoder:
```c
bool String_decodeNextWChar(MBStringDecoderState* ps);
```
The input string is specified as `ps->str` and the next character decoded will be outputted in `ps->ch`.

For the encoder:
```c
/* The typedef is to allow different encoders to coexist. I have thought of an encoder that would present all non-printable characters as "\xXXXX" escape sequence. */
typedef ATTR_NONNULL void (*EncodeWChar)(WCharEncoderState* ps, /*wchar_t or int*/ wc);
void String_encodeWChar(WCharEncoderState* ps, /*wchar_t or int*/ wc);
```
The next character that would be encoded is specified as `wc` argument here and will be appended to the end of `ps->buf`. The `WCharEncoderState` object does not otherwise know the next `wchar_t` to encode.

Perhaps the name `String_encodeWChar` is obscure on what it does. Maybe the name `WCharEncoder_appendWChar` will be better? I'm not sure yet.

BenBE commented on XUtils.c:
```diff
@@ -259,6 +260,299 @@ size_t strnlen(const char* str, size_t maxLen) {
+   return totalWidth;
+}
+
+int String_mbswidth(const char** str, size_t maxLen, int maxWidth) {
```
Why have the public API of this function modify an input argument to move to the end of the function?

Explorer09 replied:
This function has two outputs: One is the terminal width (i.e. num of terminal columns) and the other is the end character position after calculating the width.

The second output is useful to tell, for example, how many Unicode characters would it take to fit the number of terminal columns.

This functions has two uses combined in a single API: (1) to count the number of terminal columns for a given multibyte string; (2) to count the number characters it takes to fit the given terminal width.

BenBE replied:
It might seem redundant, but I'd prefer if this output parameter was a separate (NULL-able) argument to the function. This separates concerns (input argument vs. optional additional output) and makes the API clearer. This also allows to just skip the extra local variable in case you don't care AND don't have it unexpectedly be changed.

Explorer09 replied:
> It might seem redundant, but I'd prefer if this output parameter was a separate (NULL-able) argument to the function. This separates concerns (input argument vs. optional additional output) and makes the API clearer. This also allows to just skip the extra local variable in case you don't care AND don't have it unexpectedly be changed.

A `static inline` function that wraps around this `String_mbswidth()` for a more convenient output interface? That wouldn't hurt, I guess.

By the way, the "input pointer that would be modified on output" is not my original idea. There are precedents in [mbsrtowcs(3)](https://man7.org/linux/man-pages/man3/mbsrtowcs.3.html) and [mbsnrtowcs(3)](https://man7.org/linux/man-pages/man3/mbsnrtowcs.3.html) (compare mbsrtowcs(3) with [mbstowcs(3)](https://man7.org/linux/man-pages/man3/mbstowcs.3.html) and I would argue that the `const char**` argument in the former makes the API more flexible).

BenBE replied:
I gave both variants of the API to Copilot to compare (without initially disclosing what the function would do) and except for "if it runs in a loop for parsing" the additional out argument was the preferred variant. This only solidified further once I gave Copilot also some more semantics of the function it just had rated. One interesting aspect was that modern compiler tend to optimize the additional out parameter variant better. So there's several aspects that lean towards that variant of the API.

Using the C standard library's string function API as a baseline for API quality sets a very low bar for quality … ;-)

Explorer09 replied:
> I gave both variants of the API to Copilot to compare (without initially disclosing what the function would do) and except for "if it runs in a loop for parsing" the additional out argument was the preferred variant.

What variants of the API did you give to Copilot in particular?

> One interesting aspect was that modern compiler tend to optimize the additional out parameter variant better. So there's several aspects that lean towards that variant of the API.

Again I'm not sure what variant you are referring to.

> Using the C standard library's string function API as a baseline for API quality sets a very low bar for quality … ;-)

I personally value utility of the API over the "quality" which I'm not sure what it is. Convenience of passing the arguments and calling the function? Register and stack usage?

It's not I disagree with changing the API, but I didn't see an API suggestion to compare with, thus I cannot judge anything regarding what Copilot said.

Explorer09 committed:
Introduce multibyte string scanner functions

* String_makePrintable()
* EncodePrintableString()
* String_lineBreakWidth()
* String_mbswidth()

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Unicode character support in screen tab names

Compute the terminal widths of screen tab names using
String_mbswidth(). No longer assume each character takes one terminal
column.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>
